### PR TITLE
KAFKA-14032: Dequeue time for forwarded requests is ignored to set

### PR DESCRIPTION
**Pull request by yufeiyan1220**
…issues in statistics

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by showuon — COMMENTED** [review]
@yufeiyan1220 , thanks for the PR. I'm not very familiar with the request metric. But the change makes sense to me. Could you add tests for the change? Thanks.

cc @ijuma @rajinisivaram

**Comment by yufeiyan1220** [comment]
> @yufeiyan1220 , thanks for the PR. I'm not very familiar with the request metric. But the change makes sense to me. Could you add tests for the change? Thanks.
> 
> cc @ijuma @rajinisivaram

Thank you for your reply! It seems like there is no tests case for those detailed time records. In my opinion, a simulation of  dequeueing an envelope request should be made. And then we could compare the dequeue time between the envelope request and the parsed forwarded request. Is this test case make sense?

**Comment by showuon** [comment]
> Thank you for your reply! It seems like there is no tests case for those detailed time records. In my opinion, a simulation of dequeueing an envelope request should be made. And then we could compare the dequeue time between the envelope request and the parsed forwarded request. Is this test case make sense?

Make sense.

**Review by hachikuji — COMMENTED** [review]
Nice find. I had a separate fix for local time here: https://github.com/apache/kafka/commit/6bf5bfc2982158c3a1bfff4a0f65ea901ea84e7a. I think we could add a similar(ish) test case in `KafkaApisTest`.

**Comment by yufeiyan1220** [comment]
> Nice find. I had a separate fix for local time here: [6bf5bfc](https://github.com/apache/kafka/commit/6bf5bfc2982158c3a1bfff4a0f65ea901ea84e7a). I think we could add a similar(ish) test case in `KafkaApisTest`.

Thank you for your advice. I found that the only existed test is `testEnvelopeRequestWithAlterConfig` which has comparison between the envelop field and the original enveloped request after handling. So, I put the comparison logic here. Is it good for the PR or I need make it separated with this test.

**Comment by yufeiyan1220** [comment]
> Nice find. I had a separate fix for local time here: [6bf5bfc](https://github.com/apache/kafka/commit/6bf5bfc2982158c3a1bfff4a0f65ea901ea84e7a). I think we could add a similar(ish) test case in `KafkaApisTest`.

could you help me to give a review to this PR. It is ok for my change?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 1]
```diff
@@ -307,8 +307,12 @@ class KafkaApisTest {
         Seq(new AlterConfigsRequest.ConfigEntry("foo", "bar")).asJava))
     val alterConfigsRequest = new AlterConfigsRequest.Builder(configs.asJava, false).build(requestHeader.apiVersion)
 
+    val startTimeNanos = time.nanoseconds()
+    val dequeueCostNanos = 5 * 1000 * 1000
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: nit: how about `queueDurationNanos`?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 2]
```diff
@@ -321,6 +325,8 @@ class KafkaApisTest {
       any()
     )
     assertEquals(Some(request), capturedRequest.getValue.envelope)
+    // the dequeue time of forwarded request should equels to envelop request
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: nit: equels -> equals

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 3]
```diff
@@ -307,8 +307,12 @@ class KafkaApisTest {
         Seq(new AlterConfigsRequest.ConfigEntry("foo", "bar")).asJava))
     val alterConfigsRequest = new AlterConfigsRequest.Builder(configs.asJava, false).build(requestHeader.apiVersion)
 
+    val startTimeNanos = time.nanoseconds()
+    val dequeueCostNanos = 5 * 1000 * 1000
     val request = TestUtils.buildEnvelopeRequest(
-      alterConfigsRequest, kafkaPrincipalSerde, requestChannelMetrics, time.nanoseconds())
+      alterConfigsRequest, kafkaPrincipalSerde, requestChannelMetrics, startTimeNanos)
+    // add dequeue time to simulate request handlers poll request from requestQueue
+    request.requestDequeueTimeNanos = startTimeNanos + dequeueCostNanos
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Would it make sense to add `requestDequeueTimeNanos` to `buildEnvelopeRequest`?
```scala
  def buildEnvelopeRequest(
    request: AbstractRequest,
    principalSerde: KafkaPrincipalSerde,
    requestChannelMetrics: RequestChannel.Metrics,
    startTimeNanos: Long,
    dequeueTimeNanos: Long = -1,
    fromPrivilegedListener: Boolean = true
  ): RequestChannel.Request = {
```

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch! Left some minor comments.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 3 above)
Comment (yufeiyan1220) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Done

**Review by hachikuji — APPROVED** [review]
Thanks for the patch! LGTM

**Pull request merged**